infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/Workflow.java

```diff
@@ -0,0 +1,28 @@
+package org.corfudb.infrastructure.orchestrator;
+
+import java.util.List;
+
+/**
+ *
+ * A workflow is an abstract container that specifies a series of steps that achieves a
+ * multi-step operation. For example, adding a new node to the cluster, which requires
+ * several steps to complete.
+ *
+ * Created by Maithem on 10/25/17.
+ */
+
+public abstract class Workflow {
+
+    /**
+     * Return the name of this workflow
+     * @return workflow's name
+     */
+    public abstract String getName();
+
+    /**
+     * Returns the actions that are associated with
+     * this workflow.
+     * @return List of actions
+     */
+    public abstract List<Action> getActions();
```

Comment:
Seems highly limiting, can actions pass state to one another? Why return a list? Probably should return a stream.

Reply:
If actions pass state locally, what do you think should happen if one action completes, but fails to pass state to the next action?

Reply:
Also, a stream doesn't seem appropriate for the current model because you need direct access into the "stream". For example, if you fail at action 4, you want to restart the 4th action. You can process the stream and emit no-ops until you reach the required action, but that doesn't seem necessary.

Reply:
That seems easily handled by a stream...
```java
getActions().forEach(x -> {
while (true) {
    try {
x.execute();
return;
} catch (ExecutionFailedException e) {
log.info("Execution failed, retrying", e);
}
}
});
```

Not sure what no-ops are about.

Reply:
That wasn't my point. If you know that a workflow failed on step 4 out of 8 after the executing node dies, the you can pick up that workflow and continue executing from another orchestrator node, which means you would want to start from the 4th step (i.e. direct access into the 4th index of the action list). Although this can be simulated via streams by skipping the first 3, It seems that the first approach is more natural. 

A List already signifies order and random access, so I don't see a strong case for using a stream. If you strongly prefer programming that way, then I can add an extra method that returns the actions as a stream.